File: src/lib/metroCluster.ts
/**
 * Shared metro-scale clustering (#116). Geo-groups a trip's activity Locations into distinct
 * metro clusters and matches each to every covering lodging within a metro-scale radius — the one
 * detector #115 locks the optimizer's coverage check (#118), the post-import wizard's per-metro
 * prompts (#119), and #110's cross-metro warning onto, so no second heuristic re-derives this.
 *
 * Single-linkage clustering by radius, not k-means (optimizer.ts's day-clustering): there is no
 * known cluster count here, and a metro's stops should merge as one group regardless of how many
 * there are. Mirrors the same radius-growing approach transitGraphIngest.ts uses for station
 * clusters, applied to activities instead of stop nodes.
 *
 * Generic over the point shape (not pinned to the domain `Activity`/`Lodging` types) so both the
 * DB-backed callers (full `Location`s, nullable lat/lng) and the optimizer's own `LocationInput`
 * (no `kind`, non-null lat/lng) can cluster through the same code rather than each converting to
 * the other's shape first.
 */

import { haversineMeters, type Point } from "@/lib/geo";

/** Distance below which two locations count as the same metro rather than distinct destinations.
 * Wide enough to span one metro's spread (central city to its suburbs) but well under the gap
 * between actually-distant destinations (Osaka↔Tokyo, ~400km) — a build-time tunable, seeded from
 * #110's "inter-metro gap, not intra-city spread" scoping. Single source of truth: no caller
 * defines its own threshold. */
export const METRO_CLUSTER_RADIUS_METERS = 75_000;

interface Geocodable {
  lat: number | null;
  lng: number | null;
}

export interface MetroCluster<A extends Geocodable, L extends Geocodable> {
  /** May be **empty**: a lodging covering no activity-founded metro founds its own (ADR-0020,
   * amended 2026-08-17), and a metro you sleep in with nothing planned yet has no activities at
   * all. Consumers must not assume `activities[0]` exists. */
  activities: A[];
  centroid: Point;
  /** Every lodging within METRO_CLUSTER_RADIUS_METERS of the centroid — a metro can have more than
   * one covering lodging (e.g. a mid-stay hotel change), so this is never collapsed to "the"
   * lodging. Empty when no lodging in the trip reaches it. */
  lodgings: L[];
}

function pointOf<T extends Geocodable>(l: T): Point | null {
  if (l.lat == null || l.lng == null) return null;
  const p = { lat: l.lat, lng: l.lng };
  return p.lat !== 0 || p.lng !== 0 ? p : null;
}

function centroidOf(points: Point[]): Point {
  return {
    lat: points.reduce((sum, p) => sum + p.lat, 0) / points.length,
    lng: points.reduce((sum, p) => sum + p.lng, 0) / points.length,
  };
}

type Placed<T> = { item: T; point: Point };

function placed<T extends Geocodable>(items: T[]): Placed<T>[] {
  return items
    .map((item) => ({ item, point: pointOf(item) }))
    .filter((r): r is Placed<T> => r.point !== null);
}

/** Single-linkage grouping at METRO_CLUSTER_RADIUS_METERS — one member within the radius pulls the
 * whole group in, so a metro's spread merges regardless of how many members it has. Shared by both
 * founding passes below so they can't drift into two thresholds. */
function groupByProximity<T>(items: Placed<T>[]): Placed<T>[][] {
  const remaining = [...items];
  const groups: Placed<T>[][] = [];

  while (remaining.length > 0) {
    const bucket = [remaining.shift()!];
    let grew = true;
    while (grew) {
      grew = false;
      for (let i = remaining.length - 1; i >= 0; i--) {
        if (
          bucket.some(
            (b) =>
              haversineMeters(b.point, remaining[i].point) <=
              METRO_CLUSTER_RADIUS_METERS,
          )
        ) {
          bucket.push(...remaining.splice(i, 1));
          grew = true;
        }
      }
    }
    groups.push(bucket);
  }

  return groups;
}

/**
 * Groups a trip's Locations into metros (ADR-0020, amended 2026-08-17). Locations without real
 * coordinates (not yet geocoded) are dropped — they carry no geography to group on.
 *
 * Two founding passes, in this order, and the order is the point:
 *
 * 1. **Activities found metros**, each matched to every lodging within the radius of its centroid —
 *    a metro can have several (a mid-stay hotel change), so this is never collapsed to "the" lodging.
 * 2. **A lodging covering none of them founds its own**, merging with other such lodgings. A place
 *    you sleep is a destination whether or not anything is planned there yet, and without this a
 *    lodging-only region has no metro at all — which reaches `request.ts`'s "no anchor resolved"
 *    fallback and hands that Day *every* metro's skills.
 *
 * Running the second pass over what the first one left, rather than seeding one pass with both
 * kinds, is what keeps it safe: a lodging sitting between two activity groups would **bridge** them
 * into a single metro under single-linkage, silently changing coverage for trips unrelated to the
 * case this exists for. Pass 2 cannot bridge, because the groups it might have bridged are closed.
 */
export function clusterByMetro<A extends Geocodable, L extends Geocodable>(
  activities: A[],
  lodgings: L[],
): MetroCluster<A, L>[] {
  const validLodgings = placed(lodgings);

  const activityFounded = groupByProximity(placed(activities)).map((group) => {
    const centroid = centroidOf(group.map((g) => g.point));
    return {
      activities: group.map((g) => g.item),
      centroid,
      lodgings: validLodgings
        .filter(
          (l) =>
            haversineMeters(l.point, centroid) <= METRO_CLUSTER_RADIUS_METERS,
        )
        .map((l) => l.item),
    };
  });

  // Reference identity, not id: these are the very objects pass 1 just put in its `lodgings`.
  const covered = new Set(activityFounded.flatMap((m) => m.lodgings));
  const lodgingFounded = groupByProximity(
    validLodgings.filter((l) => !covered.has(l.item)),
  ).map((group) => ({
    activities: [] as A[],
    centroid: centroidOf(group.map((g) => g.point)),
    lodgings: group.map((g) => g.item),
  }));

  // Activity-founded first, so an existing trip's ordinals don't shift when a lodging-founded
  // metro appears — `preflight.ts` uses the index as the skill ordinal.
  return [...activityFounded, ...lodgingFounded];
}

File: swift/TripKrakenKit/Sources/TripKrakenKit/MetroCluster.swift
/// Shared metro-scale clustering — geo-groups a trip's activity Locations into distinct metro
/// clusters and matches each to every covering lodging within a metro-scale radius. One detector
/// locks the optimizer's coverage check, the post-import wizard's per-metro prompts, and the map's
/// metro tabs onto the same result, so no second heuristic re-derives this.
///
/// Single-linkage clustering by radius, not k-means: there is no known cluster count in advance,
/// and a metro's stops should merge as one group regardless of how many there are.
///
/// Generic over the point shape, not pinned to `Activity`/`Lodging`, so this stays testable with
/// plain fixtures the way `metroCluster.test.ts` exercises it — mirroring the TS original's own
/// generic parameterization, kept here even though this Swift codebase (ADR-0038) has only the one
/// real caller shape so far.

/// Anything that may or may not have real coordinates yet.
public protocol Geocodable {
    var lat: Double? { get }
    var lng: Double? { get }
}

extension Activity: Geocodable {
    public var lat: Double? { base.lat }
    public var lng: Double? { base.lng }
}

extension Lodging: Geocodable {
    public var lat: Double? { base.lat }
    public var lng: Double? { base.lng }
}

/// Distance below which two locations count as the same metro rather than distinct destinations.
/// Wide enough to span one metro's spread (central city to its suburbs) but well under the gap
/// between actually-distant destinations (Osaka↔Tokyo, ~400km). Single source of truth: no caller
/// defines its own threshold.
public let metroClusterRadiusMeters = 75_000.0

public struct MetroCluster<A: Geocodable, L: Geocodable> {
    /// May be **empty**: a lodging covering no activity-founded metro founds its own (ADR-0020,
    /// amended 2026-08-17), and a metro you sleep in with nothing planned yet has no activities at
    /// all. Consumers must not assume `activities.first` exists.
    public var activities: [A]
    public var centroid: Point
    /// Every lodging within `metroClusterRadiusMeters` of the centroid — a metro can have more than
    /// one covering lodging (e.g. a mid-stay hotel change), so this is never collapsed to "the"
    /// lodging. Empty when no lodging in the trip reaches it.
    public var lodgings: [L]
}

private func pointOf<T: Geocodable>(_ item: T) -> Point? {
    guard let lat = item.lat, let lng = item.lng else { return nil }
    let p = Point(lat: lat, lng: lng)
    return p.hasValidCoords ? p : nil
}

private func centroidOf(_ points: [Point]) -> Point {
    precondition(!points.isEmpty)
    return Point(
        lat: points.reduce(0) { $0 + $1.lat } / Double(points.count),
        lng: points.reduce(0) { $0 + $1.lng } / Double(points.count)
    )
}

private struct Placed<T> {
    var item: T
    var point: Point
}

private func placed<T: Geocodable>(_ items: [T]) -> [Placed<T>] {
    items.compactMap { item in pointOf(item).map { Placed(item: item, point: $0) } }
}

/// Single-linkage grouping at `metroClusterRadiusMeters` — one member within the radius pulls the
/// whole group in, so a metro's spread merges regardless of how many members it has. Shared by both
/// founding passes below so they can't drift into two thresholds.
private func groupByProximity<T>(_ items: [Placed<T>]) -> [[Placed<T>]] {
    var remaining = items
    var groups: [[Placed<T>]] = []

    while !remaining.isEmpty {
        var bucket = [remaining.removeFirst()]
        var grew = true
        while grew {
            grew = false
            for i in stride(from: remaining.count - 1, through: 0, by: -1) {
                if bucket.contains(where: { haversineMeters($0.point, remaining[i].point) <= metroClusterRadiusMeters }) {
                    bucket.append(remaining.remove(at: i))
                    grew = true
                }
            }
        }
        groups.append(bucket)
    }

    return groups
}

/// Groups a trip's Locations into metros (ADR-0020, amended 2026-08-17). Locations without real
/// coordinates (not yet geocoded) are dropped — they carry no geography to group on.
///
/// Two founding passes, in this order, and the order is the point:
///
/// 1. **Activities found metros**, each matched to every lodging within the radius of its
///    centroid — a metro can have several (a mid-stay hotel change), so this is never collapsed to
///    "the" lodging.
/// 2. **A lodging covering none of them founds its own**, merging with other such lodgings. A place
///    you sleep is a destination whether or not anything is planned there yet.
///
/// Running the second pass over what the first one left, rather than seeding one pass with both
/// kinds, is what keeps it safe: a lodging sitting between two activity groups would **bridge**
/// them into a single metro under single-linkage. Pass 2 cannot bridge, because the groups it might
/// have bridged are closed.
///
/// TS tracks "already covered" lodgings by object-reference identity, which Swift's value-type
/// structs don't have; this tracks the same set by index into `validLodgings` instead.
public func clusterByMetro<A: Geocodable, L: Geocodable>(
    activities: [A], lodgings: [L]
) -> [MetroCluster<A, L>] {
    let validLodgings = Array(placed(lodgings).enumerated())

    typealias ActivityFounded = (activities: [A], centroid: Point, lodgingIndices: [Int])

    let activityFounded: [ActivityFounded] = groupByProximity(placed(activities)).map { group in
        let centroid = centroidOf(group.map(\.point))
        let matched = validLodgings.filter { haversineMeters($0.element.point, centroid) <= metroClusterRadiusMeters }
        return (activities: group.map(\.item), centroid: centroid, lodgingIndices: matched.map(\.offset))
    }

    let coveredIndices = Set(activityFounded.flatMap(\.lodgingIndices))
    let uncoveredLodgings = validLodgings
        .filter { !coveredIndices.contains($0.offset) }
        .map(\.element)

    let lodgingFounded: [MetroCluster<A, L>] = groupByProximity(uncoveredLodgings).map { group in
        MetroCluster(activities: [], centroid: centroidOf(group.map(\.point)), lodgings: group.map(\.item))
    }

    // Activity-founded first, so an existing trip's ordinals don't shift when a lodging-founded
    // metro appears — callers use the index as a stable ordinal.
    let activityClusters = activityFounded.map { founded in
        MetroCluster<A, L>(
            activities: founded.activities, centroid: founded.centroid,
            lodgings: founded.lodgingIndices.map { validLodgings[$0].element.item }
        )
    }
    return activityClusters + lodgingFounded
}
